python/lsst/cell_coadds/_cellCoaddBuilder.py

```diff
@@ -0,0 +1,385 @@
+
+    `SingleCellCoaddBuilderTask` is intended to serve as an abstract interface
+    for various concrete implementation of coaddition algorithms via its `run`
+    method. `MultipleCellCoaddBuilderTask` is the corresponding pipeline task
+    that needs to be called from the pipeline. `MultipleCellCoaddBuilderTask`
+    must contain a concrete implementation inherited from
+    `SingleCellCoaddBuilderTask` as its base class and registered with the
+    `singleCellCoaddBuilderTaskRegistry`, say using ``@registerConfigurable``
+    decorator.
+
+    See also
+    --------
+    MultipleCellCoaddBuilderTask
+    """
+
+    ConfigClass = SingleCellCoaddBuilderConfig
+    _DefaultName = "singleCellCoaddBuilder"
+
+    def __init__(self, name: str, config: SingleCellCoaddBuilderConfig, **kwargs: Any):
+        self.config = config
+        super().__init__(name=name, config=config, **kwargs)
+
+    @abstractmethod
+    def run(
+        self,
+        inputs: Mapping[ObservationIdentifiers, Tuple[DeferredDatasetHandle, lsst.geom.Box2I]],
+        cellInfo: CellInfo,
+    ) -> pipeBase.Struct:
+        """Build a single-cell coadd
+
+        The images passed in from `MultipleCellCoaddBuilderTask` are guaranteed
+        to completely overlap the outer bounding box of the cells. Any further
+        exclusion of images based on quality assessment or other criteria
+        should be dome in this method.
+
+        Parameters
+        ----------
+        inputs: `Mapping[ObservationIdentifiers, Tuple[DeferredDatasetHandle,
+                                                       lsst.geom.Box2I]]`
+            A mapping from `lsst.cell_coadds.ObservationIdentifiers`` to a
+            tuple containing a `DeferredDatasetHandle` pointing to the input
+            image (calexp or warps) and a minimal bounding box that can be read
+            without loading the entire image.
+        cellInfo: `lsst.skymap.CellInfo`
+            An `lsst.skymap.CellInfo` object with the following
+            attributes:
+            - wcs: `lsst.afw.geom.SkyWcs`
+            - outer_bbox: `lsst.geom.Box2I`
+
+        Returns
+        -------
+        A pipeBase.Struct object with the following fields:
+        image_planes, psf and inputs.
+        """
+        raise NotImplementedError()
+
+    registry = pexConfig.makeRegistry(doc="Internal registry")
+
+
+singleCellCoaddBuilderRegistry = pexConfig.makeRegistry(doc="Registry of single cell coadd builders")
+
+
+class MultipleCellCoaddBuilderConnections(
+    pipeBase.PipelineTaskConnections,
+    dimensions=("tract", "patch", "band", "skymap"),
+    defaultTemplates={"inputCoaddName": "deep", "outputCoaddName": "deep", "warpType": "direct"},
+):
+    # Since we defer loading of images, we could take in both calexp and
+    # warps as inputs. Unless, we don't want a dependency on warps existing.
+    # The type of image will be specified in MultipleCellCoaddConfig.
```

Comment:
If we did add warps here, we'd certainly want to at least disable one of the two connections in the Connections class constructor based on config to avoid that dependency on the unused thing existing.  But at this point it's far from certain we'd want to put them in the same task anyway.

Reply:
I agree with not making warps a dependency if the coaddition algorithm is going to work only with calexps. But am I understanding correctly that the two tasks would differ trivially in their connections and nothing else? The method that computes whether an input image (calexp or warp) completely overlaps with the cell doesn't care which kind of image it is given.